### PR TITLE
✨ Remove vester contract and improve snapshot system

### LINKED ISSUE #198: [Improvement] Remove vester contract and improve snapshot system
**Description**
---
The mechanism behind the vesting system is useless, it leads to have a more complicated architecture.
We also identify that it's possible to reduce snapshot fees and vesting.

**Acceptance Criteria**
---

- [ ] New proposition is implemented
- [ ] Unit and integration tests are updated

### PR DESCRIPTION
**Pull request by Bal7hazar**
Closes #198

### CODE REVIEW

File: src/yield/library.cairo [hunk 1]
```diff
@@ -52,7 +62,15 @@ func CarbonableYielder_carbonable_offseter_address_() -> (address: felt) {
 }
 
 @storage_var
-func CarbonableYielder_carbonable_vester_address_() -> (address: felt) {
+func CarbonableYielder_payment_token_address_() -> (address: felt) {
+}
+
+@storage_var
+func CarbonableYielder_claimable_(address: felt) -> (absorption: felt) {
```
Comment (Bal7hazar) [inline src/yield/library.cairo:69]: absorption typo -> amount

File: src/yield/library.cairo [hunk 2]
```diff
@@ -128,10 +150,71 @@ namespace CarbonableYielder {
...
+        let (stored_claimable) = CarbonableYielder_claimable_.read(address);
+        let (stored_claimed) = CarbonableYielder_claimed_.read(address);
+        let (contribution) = CarbonableYielder_snapshoted_user_yielder_contribution_.read(address);
+        let (total) = CarbonableYielder_snapshoted_yielder_contribution_.read();
+        let (computed_claimable, _) = unsigned_div_rem(amount * contribution, total);
+        let claimable = stored_claimable + computed_claimable - stored_claimed;
+
+        // [Check] Overflow
+        with_attr error_message("CarbonableYielder: overflow while computing claimable") {
+            assert_in_range(claimable, 0, amount + 1);
```
Comment (Bal7hazar) [inline src/yield/library.cairo:167]: Wrong, replace by computed_claimable

**Review by Bal7hazar — COMMENTED** [review]
Review with @tekkac

File: src/yield/library.cairo (hunk 1 above)
Comment (Bal7hazar) [reply src/yield/library.cairo]: Done

File: src/yield/library.cairo [hunk 3]
```diff
@@ -128,10 +150,71 @@ namespace CarbonableYielder {
...
+        let (claimable) = claimable_of(caller);
+        with_attr error_message("CarbonableYielder: nothing to claim") {
+            assert_not_zero(claimable);
+        }
+
+        // [Effect] Add claimed value
+        let (stored_claimed) = CarbonableYielder_claimed_.read(caller);
+        let claimed = stored_claimed + claimable;
+        CarbonableYielder_claimed_.write(caller, claimed);
+
```
[the remark this answers is not in the document]
Comment (Bal7hazar) [reply src/yield/library.cairo:198]: Done

File: src/yield/library.cairo [hunk 4]
```diff
@@ -206,51 +291,50 @@ namespace CarbonableYielder {
...
         alloc_locals;
 
-        // [Check] Snapshot has been executed and vestable
+        // [Check] Snapshot has been executed and provisionable
         CarbonableYielder_assert.snapshoted();
-        CarbonableYielder_assert.vestable(total_amount);
+        CarbonableYielder_assert.provisionable();
+
+        // [Check] Amount is not null
+        let (yielder_contribution) = CarbonableYielder_snapshoted_yielder_contribution_.read();
```
[the remark this answers is not in the document]
Comment (Bal7hazar) [reply src/yield/library.cairo]: Done

File: src/yield/library.cairo [hunk 5]
```diff
@@ -183,10 +283,12 @@ namespace CarbonableYielder {
 
         // [Effect] Store period shares per users
         let (count) = CarbonableOffseter.total_user_count();
-        _snapshot_iter(index=count - 1);
+        let (amount) = CarbonableYielder_amount_.read();
+        _snapshot_iter(index=count - 1, amount=amount, total=previous_yielder_absorption);
```
Comment (Bal7hazar) [inline src/yield/library.cairo:287]: Rename `total` into `previous_yielder_contribution` for readibility

File: src/yield/library.cairo [hunk 6]
```diff
@@ -260,110 +362,41 @@ namespace CarbonableYielder {
...
-                slice_period_seconds=slice_period_seconds,
-                revocable=revocable,
-                carbonable_project_address=carbonable_project_address,
-                current_time=current_time,
-                index=index - 1,
-            );
+        // [Effect] Update user claimable if total != 0
+        if (total != 0) {
+            let (stored_claimable) = CarbonableYielder_claimable_.read(user);
+            let (new_claimable, _) = unsigned_div_rem(previous_user_contribution * amount, total);
```
Comment (Bal7hazar) [inline src/yield/library.cairo:392]: Consider to store the sum of reminders and add it to the next amount to distribute

**Review by Bal7hazar — COMMENTED** [review]
Consider to add a test that make a final snapshot with no yielder contribution (and see if claimable are safe)

**Review by tekkac — APPROVED** [review]
LGTM

**Pull request merged**